### PR TITLE
[SPARK-18445][BUILD][DOCS] Fix the markdown for `Note:`/`NOTE:`/`Note that`/`'''Note:'''` across Scala/Java API documentation

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

It seems in Scala/Java,

- `Note:`
- `NOTE:`
- `Note that`
- `'''Note:'''`
- `@note`

This PR proposes to fix those to `@note` to be consistent.

**Before**

- Scala
  

- Java
  

**After**

- Scala
  

- Java
  

## How was this patch tested?

The notes were found via

```bash
grep -r "NOTE: " . | \ # Note:|NOTE:|Note that|'''Note:'''
grep -v "// NOTE: " | \  # starting with // does not appear in API documentation.
grep -E '.scala|.java' | \ # java/scala files
grep -v Suite | \ # exclude tests
grep -v Test | \ # exclude tests
grep -e 'org.apache.spark.api.java' \ # packages appear in API documenation
-e 'org.apache.spark.api.java.function' \ # note that this is a regular expression. So actual matches were mostly `org/apache/spark/api/java/functions ...`
-e 'org.apache.spark.api.r' \
...
```

```bash
grep -r "Note that " . | \ # Note:|NOTE:|Note that|'''Note:'''
grep -v "// Note that " | \  # starting with // does not appear in API documentation.
grep -E '.scala|.java' | \ # java/scala files
grep -v Suite | \ # exclude tests
grep -v Test | \ # exclude tests
grep -e 'org.apache.spark.api.java' \ # packages appear in API documenation
-e 'org.apache.spark.api.java.function' \
-e 'org.apache.spark.api.r' \
...
```

```bash
grep -r "Note: " . | \ # Note:|NOTE:|Note that|'''Note:'''
grep -v "// Note: " | \  # starting with // does not appear in API documentation.
grep -E '.scala|.java' | \ # java/scala files
grep -v Suite | \ # exclude tests
grep -v Test | \ # exclude tests
grep -e 'org.apache.spark.api.java' \ # packages appear in API documenation
-e 'org.apache.spark.api.java.function' \
-e 'org.apache.spark.api.r' \
...
```

```bash
grep -r "'''Note:'''" . | \ # Note:|NOTE:|Note that|'''Note:'''
grep -v "// '''Note:''' " | \  # starting with // does not appear in API documentation.
grep -E '.scala|.java' | \ # java/scala files
grep -v Suite | \ # exclude tests
grep -v Test | \ # exclude tests
grep -e 'org.apache.spark.api.java' \ # packages appear in API documenation
-e 'org.apache.spark.api.java.function' \
-e 'org.apache.spark.api.r' \
...
```

And then fixed one by one comparing with API documentation/access modifiers.

After that, manually tested via `jekyll build`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
I am going to leave the images from built API doc each by each change to double-check.

**Review by sandeepmreddy — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Hi @sandeepmreddy, thanks for approving but actually it is WIP. It would be great if it is reviewed when it's complete.

**Comment by srowen** [comment]
@HyukjinKwon this is ready to go right?

**Comment by HyukjinKwon** [comment]
Ah, I couldn't check all of the java documentation yet because I have some problems with javadoc8 because of several errors (although they look existing ones). So, I am thinking of trying to use different version. (Maybe I am doing something wrong assuming it is not mentioned in docs/README.md).

Will check several times more the built documentation and also update PR description.I am also fine if you are worried of conflicts from merging others. I can resolve them.

File: core/src/main/scala/org/apache/spark/Partitioner.scala [hunk 1]
```diff
@@ -101,7 +101,7 @@ class HashPartitioner(partitions: Int) extends Partitioner {
  * A [[org.apache.spark.Partitioner]] that partitions sortable records by range into roughly
  * equal ranges. The ranges are determined by sampling the content of the RDD passed in.
  *
- * Note that the actual number of partitions created by the RangePartitioner might not be the same
+ * @note The actual number of partitions created by the RangePartitioner might not be the same
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/Partitioner.scala]: **Before**

 - Scala 


 
 - Java
  

**After**

 - Scala


 - Java

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -970,7 +970,7 @@ class SparkContext(config: SparkConf) extends Logging {
    * @param valueClass Class of the values
    * @param minPartitions Minimum number of Hadoop Splits to generate.
    *
-   * '''Note:''' Because Hadoop's RecordReader class re-uses the same Writable object for each
+   * @note Because Hadoop's RecordReader class re-uses the same Writable object for each
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: **Before**

- Scala
  

- Java
  

**After**

- Scala
  

- Java

**Review by HyukjinKwon — COMMENTED** [review]
It seems `@note` plays fine with existing formats.

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -741,7 +741,8 @@ object Unidoc {
     javacOptions in (JavaUnidoc, unidoc) := Seq(
       "-windowtitle", "Spark " + version.value.replaceAll("-SNAPSHOT", "") + " JavaDoc",
       "-public",
-      "-noqualifier", "java.lang"
+      "-noqualifier", "java.lang",
+      "-tag", """note:a:Note\:"""
```
Comment (HyukjinKwon) [inline project/SparkBuild.scala]: @srowen, this was added to deal with `@note` annotation in Java documentation. It seems `@note` for scaladoc not for javadoc. So, I had to manually deal with this.

File: pom.xml [hunk 4]
```diff
@@ -2476,6 +2476,13 @@
             <artifactId>maven-javadoc-plugin</artifactId>
             <configuration>
               <additionalparam>-Xdoclint:all -Xdoclint:-missing</additionalparam>
+              <tags>
+                <tag>
+                  <name>note</name>
+                  <placement>a</placement>
+                  <head>Note:</head>
+                </tag>
+              </tags>
```
Comment (HyukjinKwon) [inline pom.xml]: Here is the equivalent one with `SparkBuild.scala`.

**Review by srowen — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/api/java/JavaRDD.scala [hunk 5]
```diff
@@ -99,27 +99,27 @@ class JavaRDD[T](val rdd: RDD[T])(implicit val classTag: ClassTag[T])
...
-   * Note: this is NOT guaranteed to provide exactly the fraction of the count
-   * of the given [[RDD]].
-   *
    * @param withReplacement can elements be sampled multiple times (replaced when sampled out)
    * @param fraction expected size of the sample as a fraction of this RDD's size
    *  without replacement: probability that each element is chosen; fraction must be [0, 1]
    *  with replacement: expected number of times each element is chosen; fraction must be >= 0
+   *
+   * @note This is NOT guaranteed to provide exactly the fraction of the count
+   * of the given [[RDD]].
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/java/JavaRDD.scala]: - Scala
  

- Java

**Comment by HyukjinKwon** [comment]
(it was simply rebased and added 5 instances introduced in another PR by https://github.com/apache/spark/pull/15889/commits/39873dc40fbd1e62283e2dc9cc4b647ed5388a2f)

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 6]
```diff
@@ -1648,13 +1650,13 @@ class Dataset[T] private[sql](
...
-   * Note: this is NOT guaranteed to provide exactly the fraction of the count
-   * of the given [[Dataset]].
-   *
    * @param withReplacement Sample with replacement or not.
    * @param fraction Fraction of rows to generate.
    * @param seed Seed for sampling.
    *
+   * @note This is NOT guaranteed to provide exactly the fraction of the count
+   * of the given [[Dataset]].
+   *
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: - Scala
  

- Java

**Comment by HyukjinKwon** [comment]
Thank you for approving Sean!

**Comment by HyukjinKwon** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala [hunk 7]
```diff
@@ -223,7 +223,9 @@ class JavaPairRDD[K, V](val rdd: RDD[(K, V)])
   /**
    * Generic function to combine the elements for each key using a custom set of aggregation
    * functions. Turns a JavaPairRDD[(K, V)] into a result of type JavaPairRDD[(K, C)], for a
-   * "combined type" C. Note that V and C can be different -- for example, one might group an
+   * "combined type" C.
+   *
+   * @note V and C can be different -- for example, one might group an
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala]: I was skimming the changes again and noticed this. This causes this note to render pretty separately from this text, but it seems to belong with the sentence above. 

Does it cause the bullet points below to be in the note? because they also seem unrelated to this 'note'.

Could be OK but wasn't sure whether this was _really_ a stand-alone note as much as a usage of the word "note"

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 8]
```diff
@@ -1014,7 +1015,7 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
    * Output the RDD to any Hadoop-supported file system, using a Hadoop `OutputFormat` class
    * supporting the key and value types K and V in this RDD.
    *
-   * Note that, we should make sure our tasks are idempotent when speculation is enabled, i.e. do
+   * @note We should make sure our tasks are idempotent when speculation is enabled, i.e. do
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Likewise I wonder if these following sentences all go into the note? they go together.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala [hunk 9]
```diff
@@ -396,7 +396,7 @@ class JavaStreamingContext(val ssc: StreamingContext) extends Closeable {
    * Create an input stream from a queue of RDDs. In each batch,
    * it will process either one or all of the RDDs returned by the queue.
    *
-   * NOTE:
+   * @note
```
Comment (srowen) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: Likewise here just want to make sure that this actually causes the following 3 lines are in the note?

**Review by srowen — COMMENTED** [review]
A few last questions now that I skimmed it again, actually

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala (hunk 8 above)
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Ah, I just checked that It seems both

```
We should make sure our tasks are idempotent when speculation is enabled, i.e. do
not use output committer that writes data directly.
```
and
``` 
There is an example in https://issues.apache.org/jira/browse/SPARK-10063 to show the bad
result of using direct output committer with speculation enabled.
```

are relevant (it seems both are related with a direct output committer). 

Also, the original documentation also concatenates both sentences.




So. let me leave it as is if it looks okay.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Ah, I left and removed a useless comment. Yes, it seems it goes together.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala (hunk 9 above)
Comment (HyukjinKwon) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaStreamingContext.scala]: Yup, it seems keeping the original format at least.

**Before**

- Scala
  

- Java
  


**After**

- Scala
  

- Java

File: core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala [hunk 10]
```diff
@@ -234,6 +234,9 @@ class JavaPairRDD[K, V](val rdd: RDD[(K, V)])
    * In addition, users can control the partitioning of the output RDD, the serializer that is use
    * for the shuffle, and whether to perform map-side aggregation (if a mapper can produce multiple
    * items with the same key).
+   *
+   * @note V and C can be different -- for example, one might group a RDD of type (Int, Int) into
+   * an RDD of type (Int, List[Int]).
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala]: It seems fine now.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/api/java/JavaPairRDD.scala]: OK, it works alright as a standalone note. It could have stayed an inline sentence too. OK either way.

File: mllib/src/main/scala/org/apache/spark/ml/feature/PCA.scala [hunk 11]
```diff
@@ -142,8 +142,8 @@ class PCAModel private[ml] (
   /**
    * Transform a vector by computed Principal Components.
    *
-   * @note Vectors to be transformed must be the same length
-   * as the source vectors given to [[PCA.fit()]].
+   * @note Vectors to be transformed must be the same lengtt as the source vectors given
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/feature/PCA.scala]: Heh sorry @HyukjinKwon but there's now a typo here: lengtt

Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/feature/PCA.scala]: Doh! I will check the same instances. I am so sorry.

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 12]
```diff
@@ -57,15 +57,18 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
...
+   * Users provide three functions:
    *
    *  - `createCombiner`, which turns a V into a C (e.g., creates a one-element list)
    *  - `mergeValue`, to merge a V into a C (e.g., adds it to the end of a list)
    *  - `mergeCombiners`, to combine two C's into a single one.
    *
    * In addition, users can control the partitioning of the output RDD, and whether to perform
    * map-side aggregation (if a mapper can produce multiple items with the same key).
+   *
+   * @note V and C can be different -- for example, one might group a RDD of type
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: "an RDD" is correct actually. Could you make this one last fix by reverting just that change? I bet there are occurrences of "a RDD" in the docs elsewhere and you're welcome to fix them too if it makes it easier to search and replace it.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: Ah... I see. Sure. I thought "a RDD" is correct but just realised it after googling it.. Will fix and replace. I see. It should be an

**Comment by srowen** [comment]
Merged to master/2.1

**Comment by HyukjinKwon** [comment]
Thank you!!

File: core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala [hunk 13]
```diff
@@ -119,7 +119,7 @@ object SparkHadoopMapReduceWriter extends Logging {
     }
   }
 
-  /** Write a RDD partition out in a single Spark task. */
+  /** Write an RDD partition out in a single Spark task. */
```
Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Why are you changing this?

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Because `a RDD` was incorrect in English.

Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Am I missing something? I'm not familiar with English grammer, but shouldn't we use `a` when the `RDD` start with `R`?

Comment (jiangxb1987) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: I searched the code base and there are both `a RDD` and `an RDD`.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: We should replace them to `an RDD`. 

I think I found and replace them by

```bash
find . -type f -name "*.scala"  | xargs sed -i '' 's/ A RDD / An RDD /g'
find . -type f -name "*.scala"  | xargs sed -i '' 's/ A RDD\./ An RDD\./g'
find . -type f -name "*.scala"  | xargs sed -i '' 's/ a RDD / an RDD /g'
find . -type f -name "*.scala"  | xargs sed -i '' 's/ a RDD\./ an RDD\./g'
find . -type f -name "*.java"  | xargs sed -i '' 's/ A RDD / An RDD /g'
find . -type f -name "*.java"  | xargs sed -i '' 's/ A RDD\./ An RDD\./g'
find . -type f -name "*.java"  | xargs sed -i '' 's/ a RDD / an RDD /g'
find . -type f -name "*.java"  | xargs sed -i '' 's/ a RDD\./ an RDD\./g'
find . -type f -name "*.r"  | xargs sed -i '' 's/ A RDD / An RDD /g'
find . -type f -name "*.r"  | xargs sed -i '' 's/ A RDD\./ An RDD\./g'
find . -type f -name "*.r"  | xargs sed -i '' 's/ a RDD / an RDD /g'
find . -type f -name "*.r"  | xargs sed -i '' 's/ a RDD\./ an RDD\./g'
find . -type f -name "*.py"  | xargs sed -i '' 's/ A RDD / An RDD /g'
find . -type f -name "*.py"  | xargs sed -i '' 's/ A RDD\./ An RDD\./g'
find . -type f -name "*.py"  | xargs sed -i '' 's/ a RDD / an RDD /g'
find . -type f -name "*.py"  | xargs sed -i '' 's/ a RDD\./ an RDD\./g'
find . -type f -name "*.md"  | xargs sed -i '' 's/ A RDD / An RDD /g'
find . -type f -name "*.md"  | xargs sed -i '' 's/ A RDD\./ An RDD\./g'
find . -type f -name "*.md"  | xargs sed -i '' 's/ a RDD / an RDD /g'
find . -type f -name "*.md"  | xargs sed -i '' 's/ a RDD\./ an RDD\./g'
```

If there are more, we should fix. Actually, i was preparing a minor PR to fix all of them including `a X..` to `an X..`.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: It depends from how the acronym is pronounced. For example, 

An MP3 (/ɛm pi θri/)
An RPG (/ɑːr pi ʤi/)
An FBI agent (/ɛf biː aɪ/)

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: BTW, I also made this mistake and found them actually `an RDD` is correct after googling :).

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Could you point where `a RDD` is? I will try to include them in a similar way above.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Are you saying the one below

```
./R/pkg/inst/tests/testthat/test_sparkSQL.R:test_that("jsonRDD() on a RDD with json string", {
```

?

If so, I intentionally excluded this as this is the change for documentation and that was not the part of documentation. Maybe, I was trying to be too cautious. Will include them in a minor PR i am going to submit.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Yes, "an RDD" is correct because "a" precedes consonant _sounds_ and "an" precedes vowel sounds. So it's "an _arr-dee-dee_".

We don't need to fix up every last instance of this. It was a good idea to fix it all while fixing some that overlapped with this PR though.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: Ah, I will include 

```
./R/pkg/inst/tests/testthat/test_sparkSQL.R:test_that("jsonRDD() on a RDD with json string", {
./R/pkg/R/RDD.R:#' @param x A RDD.
```

It seems I should have used `-name "*.R"` instead.

Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/internal/io/SparkHadoopMapReduceWriter.scala]: I see. Sure, let me just keep it and will leave some comments in the related PRs in the future if possible.
